File: backend/app/catalog.py
import json
import re
from pathlib import Path

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import UserColor

_PATH = Path(__file__).with_name("catalog.json")
BASE: list[dict] = json.loads(_PATH.read_text(encoding="utf-8"))
BASE_BY_CODE: dict[str, dict] = {c["code"]: c for c in BASE}
BASE_CODES: frozenset[str] = frozenset(BASE_BY_CODE)


def _series_of(code: str) -> str:
    m = re.match(r"^[A-Za-z]+", code)
    return m.group(0).upper() if m else "自定义"


def effective_codes(session: Session, user_id: int) -> set[str]:
    customs = session.scalars(
        select(UserColor.code).where(
            UserColor.user_id == user_id, UserColor.source == "custom"
        )
    ).all()
    return set(BASE_CODES) | set(customs)


def effective_catalog(session: Session, user_id: int) -> list[dict]:
    rows = session.scalars(select(UserColor).where(UserColor.user_id == user_id)).all()
    overrides = {r.code: r.hex for r in rows if r.source == "override"}
    customs = [r for r in rows if r.source == "custom"]

    out = [
        {"code": c["code"], "series": c["series"], "hex": overrides.get(c["code"], c["hex"])}
        for c in BASE
    ]
    out.extend({"code": r.code, "series": _series_of(r.code), "hex": r.hex} for r in customs)
    return out

File: frontend/src/color/catalog.ts
import baseData from '../data/catalog.json';
import { hexToRgb, rgbToLab, type RGB, type Lab } from './color';

export interface BaseColor {
  code: string;
  series: string;
  hex: string;
}

export interface UserColor {
  code: string;
  hex: string;
  source: 'override' | 'custom';
  base_hex?: string;
}

export interface EffectiveColor {
  code: string;
  series: string;
  hex: string;
  rgb: RGB;
  lab: Lab;
  source: 'base' | 'override' | 'custom';
}

export function loadBaseCatalog(): BaseColor[] {
  return (baseData as BaseColor[]).map((c) => ({ ...c }));
}

export function seriesOf(code: string): string {
  const m = code.trim().match(/^[A-Za-z]+/);
  return m ? m[0].toUpperCase() : '自定义';
}

function toEffective(
  code: string,
  series: string,
  hex: string,
  source: EffectiveColor['source'],
): EffectiveColor {
  const rgb = hexToRgb(hex);
  return { code, series, hex: hex.toUpperCase(), rgb, lab: rgbToLab(rgb), source };
}

export function buildEffectiveCatalog(base: BaseColor[], userColors: UserColor[]): EffectiveColor[] {
  const overrides = new Map<string, string>();
  const customs: UserColor[] = [];
  const baseCodes = new Set(base.map((c) => c.code));

  for (const uc of userColors) {
    if (baseCodes.has(uc.code)) overrides.set(uc.code, uc.hex);
    else customs.push(uc);
  }

  const result: EffectiveColor[] = base.map((c) => {
    const overrideHex = overrides.get(c.code);
    return overrideHex
      ? toEffective(c.code, c.series, overrideHex, 'override')
      : toEffective(c.code, c.series, c.hex, 'base');
  });

  for (const uc of customs) {
    result.push(toEffective(uc.code, seriesOf(uc.code), uc.hex, 'custom'));
  }

  return result;
}
